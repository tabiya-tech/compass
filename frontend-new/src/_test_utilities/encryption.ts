import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
