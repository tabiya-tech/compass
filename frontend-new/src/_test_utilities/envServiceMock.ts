jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
