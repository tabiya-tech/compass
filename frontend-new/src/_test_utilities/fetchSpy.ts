import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

