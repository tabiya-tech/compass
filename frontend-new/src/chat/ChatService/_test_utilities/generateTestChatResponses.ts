import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
