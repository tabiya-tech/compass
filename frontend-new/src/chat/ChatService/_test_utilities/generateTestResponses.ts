import { LastMessage, RootObject } from "src/chat/ChatService/ChatService.types";

export const generateRootObjectResponse = (): RootObject => ({
  last: {
    message_for_user: "Hello",
    finished: false,
    agent_type: "WelcomeAgent",
    reasoning: "Initial greeting",
    agent_response_time_in_sec: 1.0,
    llm_stats: [
      {
        error: "",
        prompt_token_count: 50,
        response_token_count: 10,
        response_time_in_sec: 1.0,
      },
    ],
  },
  conversation_context: {
    all_history: {
      turns: [
        {
          index: 1,
          input: {
            message: "Hello",
          },
          output: {
            message_for_user: "Hi there!",
            finished: false,
            agent_type: "WelcomeAgent",
            reasoning: "Greeting response",
            agent_response_time_in_sec: 0.5,
            llm_stats: [
              {
                error: "",
                prompt_token_count: 30,
                response_token_count: 10,
                response_time_in_sec: 0.5,
              },
            ],
          },
        },
      ],
    },
    history: {
      turns: [],
    },
    summary: "",
  },
});

export const generateLastMessageResponse = (): LastMessage => ({
  message_for_user: "Hi there!",
  finished: false,
  agent_type: "WelcomeAgent",
  reasoning: "Greeting response",
  agent_response_time_in_sec: 1.0,
  llm_stats: [
    {
      error: "",
      prompt_token_count: 30,
      response_token_count: 10,
      response_time_in_sec: 0.5,
    },
  ],
});
