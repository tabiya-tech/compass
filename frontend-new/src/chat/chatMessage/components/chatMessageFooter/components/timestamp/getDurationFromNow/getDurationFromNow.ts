/**
 * Calculates and formats the duration between the given date and the current date,
 * returning a human-readable string such as "5 days ago", "10 minutes ago", or "on 10/10/2021".
 *
 * The function uses a translation function (`t`) to support localization, allowing
 * placeholders like `{{time}}` to be replaced dynamically (e.g., "5 days ago" → `t("ago", { time: "5 days" })`).
 *
 * @param {Date | string} givenDate - The past date to compare with the current time. Can be a Date object or ISO string.
 * @param {(key: string, opts?: any) => string} t - Translation function that supports interpolation via `{{time}}`.
 * @returns {string} A human-readable string representing how long ago the given date occurred.
 *
 * @example
 * // With English translation function
 * getDurationFromNow(new Date(Date.now() - 5 * 24 * 60 * 60 * 1000), t);
 * // => "5 days ago"
 *
 * getDurationFromNow("2021-10-10", t);
 * // => "on 10/10/2021"
 */
export function getDurationFromNow(givenDate: Date | string, t: any): string {
  const now = getSafeDate(new Date());
  const given = getSafeDate(givenDate);
  const duration = now.getTime() - given.getTime();

  if (duration < 0) {
    console.warn("Invalid date range: First date must be before second date", {
      now: now.toISOString(),
      given: given.toISOString(),
    });
  }

  const days = Math.floor(duration / (1000 * 60 * 60 * 24));
  const hours = Math.floor((duration % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
  const minutes = Math.floor((duration % (1000 * 60 * 60)) / (1000 * 60));

  // More than a week → format date
  if (days > 7) {
    return `on ${given.toLocaleDateString("en-GB", { timeZone: "UTC" })}`;
  }

  // Handle special and pluralized cases
  if (days === 1) return t("yesterday");
  if (days > 1)
    return t("ago", { time: `${days} ${pluralize(days, t(days === 1 ? "day" : "days"))}` });
  if (hours > 0)
    return t("ago", { time: `${hours} ${pluralize(hours, t(hours === 1 ? "hour" : "hours"))}` });
  if (minutes > 0)
    return t("ago", { time: `${minutes} ${pluralize(minutes, t(minutes === 1 ? "minute" : "minutes"))}` });

  return t("just_now");
}

/**
 * Returns the pluralized form of a time unit (e.g., "day" vs. "days").
 *
 * Note: In English, this simply adds or omits the "s" suffix. For other languages,
 * you should modify the translation keys instead of this logic.
 *
 * @param {number} value - The numeric amount of the unit.
 * @param {string} unit - The base unit (e.g., "day", "hour", "minute").
 * @returns {string} The pluralized or singular form of the unit.
 *
 * @example
 * pluralize(1, "day");   // => "day"
 * pluralize(3, "hour");  // => "hour"
 */
function pluralize(value: number, unit: string): string {
  return value === 1 ? unit : `${unit}`;
}

/**
 * Ensures the input is a valid Date object and normalizes it to UTC.
 * Accepts both Date objects and date strings.
 *
 * @param {Date | string} date - The date input to normalize.
 * @returns {Date} A safe Date object in UTC.
 *
 * @example
 * getSafeDate("2021-10-10");
 * // => Date object representing 2021-10-10T00:00:00.000Z
 */
function getSafeDate(date: Date | string): Date {
  return date instanceof Date ? new Date(date.toISOString()) : new Date(new Date(date).toISOString());
}
