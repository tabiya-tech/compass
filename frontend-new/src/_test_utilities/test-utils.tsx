// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
