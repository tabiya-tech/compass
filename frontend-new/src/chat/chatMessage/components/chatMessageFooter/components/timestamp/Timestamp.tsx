import { useTranslation } from "react-i18next";
import { Typography, styled } from "@mui/material";
import { getDurationFromNow } from "src/chat/chatMessage/components/chatMessageFooter/components/timestamp/getDurationFromNow/getDurationFromNow";

const uniqueId = "c9253326-05ee-4bff-9d43-7852ca78a033";

export const DATA_TEST_ID = {
  TIMESTAMP: `chat-message-timestamp-${uniqueId}`,
};

export interface ChatMessageFooterProps {
  sentAt: string
}

const TimeStamp = styled(Typography)(({ theme }) => ({
  fontSize: theme.typography.body2.fontSize,
  color: theme.palette.text.secondary,
  marginTop: theme.spacing(theme.tabiyaSpacing.xs),
}));


const Timestamp: React.FC<ChatMessageFooterProps> = ({ sentAt }) => {
  const { t } = useTranslation();
  let duration;
  try {
    duration = getDurationFromNow(new Date(sentAt),t);
  } catch (e) {
    console.error(new Error("Failed to get message duration", { cause: e }));
  }

  return (
    <TimeStamp data-testid={DATA_TEST_ID.TIMESTAMP} variant="caption">
      {t("sent")} {duration}
    </TimeStamp>
  )
}

export default Timestamp;