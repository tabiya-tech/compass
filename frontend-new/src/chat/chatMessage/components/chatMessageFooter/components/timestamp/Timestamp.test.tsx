// mute the console
import "src/_test_utilities/consoleMock";

import Timestamp , { DATA_TEST_ID } from "src/chat/chatMessage/components/chatMessageFooter/components/timestamp/Timestamp";
import { render, screen } from "src/_test_utilities/test-utils";
import { getDurationFromNow } from "src/chat/chatMessage/components/chatMessageFooter/components/timestamp/getDurationFromNow/getDurationFromNow";

jest.mock("src/chat/chatMessage/components/chatMessageFooter/components/timestamp/getDurationFromNow/getDurationFromNow", () => {
  return {
    getDurationFromNow: jest.fn()
  }
})

describe("render tests", () => {
  test("should render the Chat Message Timestamp without a child if none is passed", () => {
    // GIVEN a sent at time
    const givenSentAt: string = new Date().toISOString()

    // AND a getDurationFromNow function that returns a given duration string
    const givenDuration = "a foo years ago";
    (getDurationFromNow as jest.Mock).mockReturnValueOnce(givenDuration);

    // WHEN the Chat Message Timestamp is rendered
    render(<Timestamp sentAt={givenSentAt} />);

    // THEN expect the timestamp to be visible
    expect(screen.getByTestId(DATA_TEST_ID.TIMESTAMP)).toBeInTheDocument();
    // AND expect the getDurationFromNow method to have been called with the given sentAt as a Date object
    expect(getDurationFromNow).toHaveBeenCalledWith(new Date(givenSentAt));
    // AND expect the given duration to be shown
    expect(screen.getByTestId(DATA_TEST_ID.TIMESTAMP)).toHaveTextContent(givenDuration);

    // AND expect the component to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.TIMESTAMP)).toMatchSnapshot();
  });
});