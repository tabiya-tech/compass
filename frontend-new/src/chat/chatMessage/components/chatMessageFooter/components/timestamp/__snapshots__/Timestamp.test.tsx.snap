// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render the Chat Message Timestamp without a child if none is passed 1`] = `
<span
  class="MuiTypography-root MuiTypography-caption css-1l3yql3-MuiTypography-root"
  data-testid="chat-message-timestamp-c9253326-05ee-4bff-9d43-7852ca78a033"
>
  sent
   
  a foo years ago
</span>
`;
